File: Terceiro semestre/POO/runcodes/2/Tabuleiro.py
class Tabuleiro():
    
    def __init__(self,  tamanho_matriz, matriz_dada) -> None:
        
        self.tamanho_matriz = tamanho_matriz   # Toda mudança em matriz vai acontecer em matriz_dada e vice-versa
        self.matriz = matriz_dada
        self.lista_posicao_vazia = list()
        for i in range(tamanho_matriz):
            for j in range(tamanho_matriz):
                if(self.matriz[i][j] == 0):
                    self.lista_posicao_vazia.append(i)
                    self.lista_posicao_vazia.append(j)
                    return
                
    
    def print_matriz(self):
        for p in range(self.tamanho_matriz):
            print("+------", end='')
            
        print("+", end='')
        
        for i in range(self.tamanho_matriz):
            print("")
            
            for j in range(self.tamanho_matriz):
                if self.matriz[i][j] == 0:
                    print("|      ", end='')
                else:
                    # Usa formatação de strings para garantir que outros números ocupem dois espaços, com espaço extra à esquerda se necessário
                    print(f"|  {self.matriz[i][j]: >2}  ", end='')

        
            print("|")
            
            for p in range(self.tamanho_matriz):
                print("+------", end='')
                
            print("+", end='')
        
        print("")
        
        
    def swap(self, linha1, coluna1, linha2, coluna2):
        self.matriz[linha1][coluna1], self.matriz[linha2][coluna2] = self.matriz[linha2][coluna2], self.matriz[linha1][coluna1]
    
    
    def up(self):
        if(self.lista_posicao_vazia[0]+1 >= self.tamanho_matriz):
            return
        self.swap(self.lista_posicao_vazia[0], self.lista_posicao_vazia[1], self.lista_posicao_vazia[0]+1, self.lista_posicao_vazia[1])

        self.lista_posicao_vazia[0] += 1
        

    def down(self):
        if(self.lista_posicao_vazia[0]-1 < 0 ):
            return
        
        self.swap(self.lista_posicao_vazia[0], self.lista_posicao_vazia[1], self.lista_posicao_vazia[0]-1, self.lista_posicao_vazia[1])

        self.lista_posicao_vazia[0] -= 1
        

    def left(self):
        if(self.lista_posicao_vazia[1]+1 >= self.tamanho_matriz):
            return
        
        self.swap(self.lista_posicao_vazia[0], self.lista_posicao_vazia[1], self.lista_posicao_vazia[0], self.lista_posicao_vazia[1]+1)
        self.lista_posicao_vazia[1] += 1
        
    

    def right(self):

        if(self.lista_posicao_vazia[1]-1 < 0 ):
            return
        

        self.swap(self.lista_posicao_vazia[0], self.lista_posicao_vazia[1], self.lista_posicao_vazia[0], self.lista_posicao_vazia[1]-1)

        self.lista_posicao_vazia[1] -= 1
        
    
    def end_game(self):
        k = 0
        for i in range(self.tamanho_matriz):
            for j in range(self.tamanho_matriz):
                if(self.matriz[i][j] != k):
                    return False
                
                k+=1
        
        return True
                
    

File: Terceiro semestre/POO/runcodes/1/Tabuleiro.java
public class Tabuleiro{

    public int[][] matriz;  
    public int[] lista_posicao_vazia;
    public int tamanho_matriz;

    public Tabuleiro(int tamanho_matriz, int [][] matriz_dada){

        this.tamanho_matriz = tamanho_matriz;

        matriz = new int [tamanho_matriz][tamanho_matriz];

        lista_posicao_vazia = new int [2];

        for(int i = 0; i < tamanho_matriz; i++){
            for(int j = 0; j < tamanho_matriz; j++){
                this.matriz[i][j] = matriz_dada[i][j];
            }
        }

        for(int i=0; i< tamanho_matriz; i++){
            for(int j=0; j<tamanho_matriz; j++){
                if(matriz[i][j] == 0){
                    lista_posicao_vazia[0] = i;
                    lista_posicao_vazia[1] = j;
                    return;
                }
            }
        }

    }

    public void print_matrix() { 
        for(int p = 0; p<tamanho_matriz; p++){
            System.out.print("+------");
        }
        
        System.out.printf("+");

        for (int i = 0; i < tamanho_matriz; i++) {

            System.out.println();

            for (int j = 0; j < tamanho_matriz; j++) {
                System.out.printf("|  %2d  ", matriz[i][j]);
            }

            System.out.printf("|");

            System.out.println();

            for(int p = 0; p<tamanho_matriz; p++){
                System.out.print("+------");
            }

            System.out.printf("+");

        }

        System.out.println();

    }

    private void swap(int linha1, int coluna1, int linha2, int coluna2){
        int tmp = matriz[linha1][coluna1];
        matriz[linha1][coluna1] = matriz[linha2][coluna2];
        matriz[linha2][coluna2] = tmp;
        
    }


    public void up(){
        if((int)lista_posicao_vazia[0]+1 >= tamanho_matriz){
            // print_matrix();
            return;
        }
        swap(lista_posicao_vazia[0], lista_posicao_vazia[1], lista_posicao_vazia[0]+1, lista_posicao_vazia[1]);

        lista_posicao_vazia[0]++;
        
    }

    public void down(){
        if((int)lista_posicao_vazia[0]-1 < 0 ){
            // print_matrix();
            return;
        }
        swap(lista_posicao_vazia[0], lista_posicao_vazia[1], lista_posicao_vazia[0]-1, lista_posicao_vazia[1]);

        lista_posicao_vazia[0]--;
        
    }

    public void left(){
        if((int)lista_posicao_vazia[1]+1 >= tamanho_matriz){
            // print_matrix();
            return;
        }
        swap(lista_posicao_vazia[0], lista_posicao_vazia[1], lista_posicao_vazia[0], lista_posicao_vazia[1]+1);
        lista_posicao_vazia[1]++;
        
    }

    public void right(){

        if((int)lista_posicao_vazia[1]-1 < 0 ){
            // print_matrix();
            return;
        }

        swap(lista_posicao_vazia[0], lista_posicao_vazia[1], lista_posicao_vazia[0], lista_posicao_vazia[1]-1);

        lista_posicao_vazia[1]--;
        
    }


    public boolean end_game(){
        for(int i =0, k = 0; i< tamanho_matriz; i++){
            for(int j=0; j< tamanho_matriz; j++, k++){
                if(matriz[i][j] != k){
                    return false;
                }
            }
        }
        return true;
    }



}




// +------+------+------+